### PR TITLE
[SPARK-18519][SQL] map type can not be used in EqualTo

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Technically map type is not orderable, but can be used in equality comparison. However, due to the limitation of the current implementation, map type can't be used in equality comparison so that it can't be join key or grouping key.

This PR makes this limitation explicit, to avoid wrong result.

## How was this patch tested?

updated tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 1]
```diff
@@ -183,21 +183,6 @@ trait CheckAnalysis extends PredicateHelper {
               s"join condition '${condition.sql}' " +
                 s"of type ${condition.dataType.simpleString} is not a boolean.")
 
-          case j @ Join(_, _, _, Some(condition)) =>
-            def checkValidJoinConditionExprs(expr: Expression): Unit = expr match {
-              case p: Predicate =>
-                p.asInstanceOf[Expression].children.foreach(checkValidJoinConditionExprs)
-              case e if e.dataType.isInstanceOf[BinaryType] =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: This doesn't make sense, `BinaryType` is comparable, we should accept it as join key.

**Comment by cloud-fan** [comment]
cc @yhuai @hvanhovell  @gatorsmile

**Comment by rxin** [comment]
So technically I think map type should be able to be used in equality comparison, but not orderable, so I'm not sure if this is correct.

Spark SQL currently has a restriction that we just don't allow map types to be used in equality comparison at all, but I don't want to structure code in a way that would make it harder to support that in the future.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 2]
```diff
@@ -388,6 +388,21 @@ abstract class BinaryComparison extends BinaryOperator with Predicate {
       defineCodeGen(ctx, ev, (c1, c2) => s"${ctx.genComp(left.dataType, c1, c2)} $symbol 0")
     }
   }
+
+  override def inputType: AbstractDataType = AnyDataType
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: isn't this "wrong"? map type is not supported here is it?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: We need this to satisfy `BinaryOperator`, the type checking logic is defined by both `BinaryOperator` and `checkInputDataTypes` here. Shall we make `BinaryComparison` not extends `BinaryOperator` and duplicate its type checking logic here?

**Comment by cloud-fan** [comment]
> Spark SQL currently has a restriction that we just don't allow map types to be used in equality comparison at all, but I don't want to structure code in a way that would make it harder to support that in the future.

If we wanna support equality comparison for map type in the future, we can just update the type checking logic for `BinaryComparison`. For now I think it's better to forbid users to do this, to avoid confusing behaviors, e.g. `SELECT MAP(1,2) = MAP(1,2)` returns false.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 3]
```diff
@@ -388,6 +388,26 @@ abstract class BinaryComparison extends BinaryOperator with Predicate {
...
+  // AnyDataType here to bypass this check, and implement the actual type checking logic in
+  // `checkInputDataTypes`
+  override def inputType: AbstractDataType = AnyDataType
+
+  // TODO: although map type is not orderable, technically map type should be able to be used
+  // in equality comparison, update this type check once we support it.
+  override def checkInputDataTypes(): TypeCheckResult = {
+    super.checkInputDataTypes() match {
+      case TypeCheckResult.TypeCheckSuccess =>
+        if (RowOrdering.isOrderable(left.dataType)) {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: Semantically, this is different from the previous lilmits:
```
override def inputType: AbstractDataType = TypeCollection.Ordered
```

We add the support of `StructType`, `ArrayType` and `UDT`. It sounds like no test case coverage. Could you please add them?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 4]
```diff
@@ -388,6 +388,26 @@ abstract class BinaryComparison extends BinaryOperator with Predicate {
       defineCodeGen(ctx, ev, (c1, c2) => s"${ctx.genComp(left.dataType, c1, c2)} $symbol 0")
     }
   }
+
+  // The type checking logic of BinaryComparison can't be expressed by `inputType`, so we use
+  // AnyDataType here to bypass this check, and implement the actual type checking logic in
+  // `checkInputDataTypes`
+  override def inputType: AbstractDataType = AnyDataType
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: hmm, Hive only supports atomic type, shall we follow it?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: FYI, postgres supports binary comparison on array type.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: let me minimize the size of this PR, I'll keep `GreaterThan`, `LessThan`, etc. unchanged, and only update `EqualTo`, `EqualNullSafe` to forbid map types.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 1 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: We should add a few tests with binary keys, that can also be done in a follow-up. The parser supports binary literals, so we can use `SQLQueryTestSuite` for this.

**Review by hvanhovell — COMMENTED** [review]
A few minor things, otherwise LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: I already added one: https://github.com/apache/spark/pull/15956/files#diff-999f50ed13c690eaec243e3b3446af08R468 , yea we should add more end-to-end tests in `SQLQueryTestSuite` in follow-up

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 5]
```diff
@@ -440,6 +463,29 @@ case class EqualNullSafe(left: Expression, right: Expression) extends BinaryComp
 
   override def inputType: AbstractDataType = AnyDataType
 
+  override def checkInputDataTypes(): TypeCheckResult = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: well it's only duplicated twice and will be removed after https://github.com/apache/spark/pull/15970 , I think it should be fine.

**Comment by hvanhovell** [comment]
LGTM pending Jenkins

**Comment by hvanhovell** [comment]
LGTM - merging to master/2.0. Thanks!

**Comment by hvanhovell** [comment]
@cloud-fan I cannot merge this to 2.0. Can you make a backport if we need one?

**Comment by cloud-fan** [comment]
Yea, I'll backport

**Pull request closed without merging**